### PR TITLE
refactor(mobile): add Isar DB & Store class

### PR DESCRIPTION
**Pull request by fyfrey**
- add Isar as a dependency
- integrate Isar DB as a provider
- new, custom Store: globally accessible key-value store like Hive (but based on Isar)
- replace first few places of Hive usage with the new Store: assetEtag and remoteUserId, both used to fetch assets in the AssetService
- automatic migration from Hive to Store (at app startup) for assetEtag & remoteUserId
- fixed assetETag being incorrectly overwritten with null, resulting in unnecessary loading all assets from server
- replaced one AssetResponseDto usage with Asset

I think this is the last refactoring PR before actually adding offline support in the first feature PR of this series :-)

### CODE REVIEW

**Review by alextran1502 — COMMENTED** [review]
Thank you for the PR Fynn, looks good! I add a few comments.

So whenever we add a new key to the KeyStore enum, which command will we need to run to regenerate the autogenerated file? Can you put that into the ReadMe file of the mobile app project?

**Review by alextran1502 — COMMENTED** [review]
Tested on iOS and everything looks good

File: mobile/lib/shared/models/store.dart [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            : (key._fromJson == null ? value : json.encode(value.toJson())),
+      );
+}
+
+/// Key for each possible value in the `Store`.
+/// Defines the data type (int, String, JSON) for each value
+enum StoreKey {
+  userRemoteId(0),
+  assetETag(1),
+  ;
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/shared/models/store.dart:89]: correct!

File: mobile/lib/shared/providers/db.provider.dart [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:isar/isar.dart';
+
+// overwritten in main.dart due to async loading
+final dbProvider = Provider<Isar>((_) => throw UnimplementedError());
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/shared/providers/db.provider.dart:5]: I thought that, too! But it does not work with how we use providers in the app.
It turns out, you cannot `ref.watch(futureProvider)` and assign it to a usable value

I guess, we'd need to change all providers (that will depending on the new dbProvider) to be a FutureProvider as well and that changes how we'd need to use it in the code.  I guess we can do that in the future, if we want to. For now, this workaround is simpler.

File: mobile/lib/shared/services/asset.service.dart [hunk 3]
```diff
@@ -37,15 +38,15 @@ class AssetService {
       final Pair<List<AssetResponseDto>, String?>? remote =
           await _apiService.assetApi.getAllAssetsWithETag(eTag: etag);
       if (remote == null) {
-        return const Pair(null, null);
+        return Pair(null, etag);
```
[the remark this answers is not in the document]
Comment (zoodyy) [reply mobile/lib/shared/services/asset.service.dart:41]: It's a bug / not optimal: Currently, an etag of null is returned if the response from the server did not change. This leads to storing null as the current etag on the client. Next time it has to request without etag (thus no chance of caching).

**Comment by zoodyy** [comment]
@alextran1502 Thanks for reviewing! And testing on iOS :-) also works for Matthias.
From my side, you can merge if you're satisfied with the answers

**Comment by zoodyy** [comment]
Added a tiny change: maximum size of the database is now limited to 256MB instead of default 1024MB. This reportedly solves issues on older iOS devices. The DB size is should still be large enough for all our data by a wide margin :-)

**Pull request merged**